### PR TITLE
Add broadcast_like.

### LINKED ISSUE #2885: Add broadcast_like?
What do you think of adding `da.broadcast_like(a)`?

This would be equivalent to `xr.broadcast(a, da)[1]`

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #2885
 - [x] Tests added
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by shoyer** [comment]
In principle, it feels like it would be better to have both `xarray.broadcast()` and `broadcast_like` written in terms of a lower level helper method that broadcasts only one object. This is what we do for `align`/`reindex_like`, for example. But in practice this is certainly way easier to write and probably good enough...

File: xarray/core/alignment.py [hunk 1]
```diff
@@ -391,6 +391,58 @@ def reindex_variables(
...
+                # ignore dim not in var.dims
+                var_dims_map[dim] = var.shape[var.dims.index(dim)]
+
+        return var.set_dims(var_dims_map)
+
+    def _broadcast_array(array):
+        data = _set_dims(array.variable)
+        coords = OrderedDict(array.coords)
+        coords.update(common_coords)
+        return DataArray(data, coords, data.dims, name=array.name,
```
Comment (max-sixty) [inline xarray/core/alignment.py]: Could potentially use `._replace`, and then only supply the changed items (but not a big deal)

Comment (dcherian) [reply xarray/core/alignment.py]: Not sure what you mean

Comment (max-sixty) [reply xarray/core/alignment.py]: i.e. `return array._replace(data, coords)`
So `array` keeps its type if someone inherited from `DataArray` and we're not coupled to specific DataArray properties if those ever change

**Comment by shoyer** [comment]
I'm merging this now, so @DavidMertz can put his doc/test improvements on top of it.

File: xarray/core/alignment.py (hunk 1 above)
Comment (shoyer) [reply xarray/core/alignment.py]: I agree, but let's save this for a later refactor.

**Comment by shoyer** [comment]
thanks @dcherian !

**Comment by dcherian** [comment]
Sounds good. @DavidMertz's docs/tests are much better!

**Pull request merged**